src/hostapi/sndio/pa_sndio.c

```diff
@@ -0,0 +1,763 @@
+            }
+            sndioStream->rpos += sndioStream->par.round;
+            timeinfo.inputBufferAdcTime = (double)sndioStream->realpos / sndioStream->par.rate;
+        }
+        if( sndioStream->mode & SIO_PLAY )
+        {
+            timeinfo.outputBufferDacTime =
+                (double)( sndioStream->realpos + sndioStream->par.bufsz ) / sndioStream->par.rate;
+        }
+        timeinfo.currentTime = sndioStream->realpos / (double)sndioStream->par.rate;
+        PaUtil_BeginBufferProcessing( &sndioStream->bufferProcessor, &timeinfo, 0 );
+        if( sndioStream->mode & SIO_PLAY )
+        {
+            PaUtil_SetOutputFrameCount( &sndioStream->bufferProcessor, sndioStream->par.round );
+            PaUtil_SetInterleavedOutputChannels( &sndioStream->bufferProcessor, 0, sndioStream->wbuf,
+                                                 sndioStream->par.pchan );
+        }
+        if( sndioStream->mode & SIO_REC )
+        {
+            PaUtil_SetInputFrameCount( &sndioStream->bufferProcessor, sndioStream->par.round );
+            PaUtil_SetInterleavedInputChannels( &sndioStream->bufferProcessor, 0, sndioStream->rbuf,
+                                                sndioStream->par.rchan );
+        }
+        result = paContinue;
+        n = PaUtil_EndBufferProcessing( &sndioStream->bufferProcessor, &result );
+        if( n != sndioStream->par.round )
+        {
+            PA_DEBUG( ( "sndioThread: %d < %u frames, result = %d\n", n, sndioStream->par.round, result ) );
+        }
+        if( result != paContinue )
+        {
+            break;
+        }
+        if( sndioStream->mode & SIO_PLAY )
+        {
+            n = sio_write( sndioStream->hdl, sndioStream->wbuf, wblksz );
+            if( n < wblksz )
+            {
+                PA_DEBUG( ( "sndioThread: sio_write failed\n" ) );
+                goto failed;
+            }
+            sndioStream->wpos += sndioStream->par.round;
+        }
+    }
+failed:
+    sndioStream->active = 0;
+    PA_DEBUG( ( "sndioThread: done\n" ) );
+    return NULL;
+}
+
+static PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi, PaStream **paStream,
+                           const PaStreamParameters *inputPar, const PaStreamParameters *outputPar, double sampleRate,
+                           unsigned long framesPerBuffer, PaStreamFlags streamFlags, PaStreamCallback *streamCallback,
+                           void *userData )
+{
+    PaSndioHostApiRepresentation *sndioHostApi = (PaSndioHostApiRepresentation *)hostApi;
+    PaSndioStream *sndioStream;
+    PaError err;
+    struct sio_hdl *hdl;
+    struct sio_par par;
+    unsigned mode;
+    int inputChannelCount, outputChannelCount;
+    PaSampleFormat ifmt, ofmt, siofmt;
+    const char *dev;
+
+    PA_DEBUG( ( "OpenStream:\n" ) );
+
+    mode = 0;
+    inputChannelCount = outputChannelCount = 0;
+    ifmt = ofmt = 0;
+    sio_initpar( &par );
+
+    if( outputPar && outputPar->channelCount > 0 )
```

Comment:
I'm stumped. Is this an outputParticle, outputParagraph, outputParallel? ;-)

Reply:
I believe to be Parameters in this context.

Reply:
Ah! Obvious in hindsight. I was honestly stumped. Thanks for the clarification.